### PR TITLE
Support mapping to fields other than id

### PR DESCRIPTION
**Pull request by TuckerWhitehouse**
Currently [mapping is hard-coded to id](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/schema/infer-graphql-type.js#L147), so this change is to enable mapping to any field.
e.g. `'MarkdownRemark.frontmatter.author': 'MarkdownRemark.frontmatter.id'`

### CODE REVIEW

**Comment by pieh** [comment]
@TuckerWhitehouse this is great!

There are some edge cases that would require more thought:
 - what if field we are trying to map to isn't unique - should we map to first instance that matches - like in your current implementation or should we return array with all matches?
 - what if we would like to match arrays?

Those doesn't need to be implemented immediately, but we need to figure them out conceptually before expanding api to not break that api later if we add more features.

Do you feel like You could add tests for this (and actually for mapping at all - as it seems those were never written) - https://github.com/TuckerWhitehouse/gatsby/blob/master/packages/gatsby/src/schema/__tests__/infer-graphql-type-test.js#L260 ? If not, don't worry - I'll handle adding tests then.

**Comment by TuckerWhitehouse** [comment]
Hey @pieh, thanks!

Based on the current implementation where mapping is one-to-one, maybe it would be appropriate to just print a warning if more than once match is found, and continue to return the first?

What do you think about a syntax like `[MarkdownRemark.frontmatter.id]` for mapping to an array? Alternatively, an options object could be used, but in either case I think it could be done in a backwards compatible way.

As far as tests, I'd be happy to add them, but the lack of any existing tests for this left me unsure of what the tests would look like.

**Comment by pieh** [comment]
I think you are right that we should return first node that matches to keep things somewhat consistent. This would need extra options to declare if you want single mapped node or if you want all matching nodes (probably via options object as you later suggested)

With arrays, I didn't explained myself - what I meant was about cases like somewhere in the path to field we want to map on there is array:
`'MarkdownRemark.frontmatter.someField': 'MarkdownRemark.frontmatter.arrayOfObjects.id'` (this won't work as path for `_.get`). 

I'll deal with tests then, don't worry about them. Will try to add them tomorrow.

**Comment by TuckerWhitehouse** [comment]
Ah, so I thought you were addressing the `one-to-one` vs `one-to-many`, but your saying gatsby may want to support mapping to a value within a complex data type.

Would this even need to be an option, or could it just be functionality that is added in the future - if the user is providing a path that includes `arrayOfObjects.id`, I feel like that's their way of opting in to that sort of behavior.

Did you have a preference for an options object vs the proposed syntax for indicating if you want only the first match or all matches? My thought with the syntax (`[MarkdownRemark.frontmatter.id]`) is that it keeps the mapping config pretty concise, and is similar to how you declare an array with [GraphQL Schema Language](http://graphql.org/learn/schema/).

**Comment by pieh** [comment]
It totally could be added in the future - as I said in first comment - this doesn't have to implemented now (or maybe ever), just trying to be as sure as possible we don't introduce any api that could potentially limit future features (i.e. if we would want to change api - we would need to wait for next major release when we can introduce breaking changes).

I personally I'm in favor of using options object for flags like "we want array" but we can do both - single string "short" syntax that gets converted to options object internally. This is not graphql schema definition so I wouldn't use that as a reason to use that notation (to be clear I'm not against using that - just against that reason :) ). 

BTW, I just remembered that I discussed briefly about feature in this PR in one of the issues ( https://github.com/gatsbyjs/gatsby/issues/3129#issuecomment-360932606 )

As far as this PR is concerned - it's only missing tests that I will try to add tomorrow and this will be good to go. I just brought up edge cases for discussion.

File: packages/gatsby/src/schema/infer-graphql-type.js [hunk 1]
```diff
@@ -132,21 +132,31 @@ function inferFromMapping(
...
-      n => n.internal.type === linkedType && _.get(n, linkedField) === fieldValue
+      n =>
+        n.internal.type === linkedType && _.get(n, linkedField) === fieldValue
     )
+    if (linkedNodes.length > 1) {
+      console.log(
+        `Multiple matching nodes found for "${
+          mapping[fieldSelector]
+        }", returning the first match`
+      )
```
Comment (pieh) [inline packages/gatsby/src/schema/infer-graphql-type.js]: This potentially could spam a lot. I think this should log only once for each mapping

Comment (TuckerWhitehouse) [reply packages/gatsby/src/schema/infer-graphql-type.js]: Yeah, I'm not sure it adds a huge amount of value anyway -- thinking it may make more sense to add a note to docs that only the first match is returned instead.

It doesn't look like `debug` is used very heavily in the code base, but would this maybe be a good use case for it? Not sure if this would actually provide enough info to help people identify why it's matching multiple documents.
```js
const debug = require(`debug`)(`gatsby:schema:infer-graphql-type`)
...
debug(`Multiple matching nodes found for ${fieldSelector} > ${mapping[fieldSelector]}`)
```

Comment (pieh) [reply packages/gatsby/src/schema/infer-graphql-type.js]: Yeah, let's just skip this log/warning for now and update docs instead. We can always revisit this later.

**Comment by KyleAMathews** [comment]
Great work both of you! Agree adding a way so you could you want an array of matched results would be a great follow-up feature. Though if we get into that then we have to think about if we want to support filtering & sorting.

**Comment by TuckerWhitehouse** [comment]
This is awesome! Thanks @pieh & @KyleAMathews, our site is already deployed to production using this :D

**Comment by jlengstorf** [comment]
Hiya @TuckerWhitehouse! 👋

This is definitely late, but on behalf of the entire Gatsby community, I wanted to say thank you for being here.

Gatsby is built by awesome people like you. Let us say “thanks” in two ways:

1.  **We’d like to send you some Gatsby swag.** As a token of our appreciation, you can go to the [Gatsby Swag Store][store] and log in with your GitHub account to get a coupon code good for one free piece of swag. (We’ve got t-shirts and hats, plus some socks that are really razzing our berries right now.)
2.  **If you’re not already part of it, we just invited you to join the Gatsby organization on GitHub.** This will add you to our team of maintainers. You’ll receive an email shortly asking you to confirm. By joining the team, you’ll be able to label issues, review pull requests, and merge approved pull requests.

If you have questions, please don’t hesitate to reach out to us: tweet at [@gatsbyjs][twitter] and we’ll come a-runnin’.

Thanks again! 💪💜

[store]: https://store.gatsbyjs.org
[twitter]: https://twitter.com/gatsbyjs
